improve reqwest backend

seanmonstar opened the pull request:
- use a single Client, yay connection pooling
- fix proxies (scheme was missing)
- fix reqwest partial downloads
- add tests (including partial downloads)
- add warning about RUSTUP_USE_HYPER var

seanmonstar committed:
improve reqwest backend

- use a single Client, yay connection pooling
- fix proxies (scheme was missing)
- fix reqwest partial downloads
- add tests (including partial downloads)
- add warning about RUSTUP_USE_HYPER var

alexcrichton commented on src/download/tests/download-curl-resume.rs:
```diff
@@ -62,36 +36,48 @@ fn partially_downloaded_file_gets_resumed_from_byte_offset() {
                                   &target_path,
                                   true,
                                   Some(&|msg| {
+        println!("event: {:?}", msg);
```
Stray println?

seanmonstar replied:
Woops, yep. Used to learn that I needed to add a `Seek` to the reqwest backend file downloader.

alexcrichton commented:
Looks great to me, thanks!

seanmonstar commented:
I noticed in using this that it would warn about the environment variable several times a run, as it tried to download several files. So I've added a static bool to only emit the warning once...

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 629aec9 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 629aec9548a1ffe3ea579b83c5a5993f636bb058 -->

bors commented:
:hourglass: Testing commit 629aec9548a1ffe3ea579b83c5a5993f636bb058 with merge bd93415a8bc1c924fd3ca07bccb49195668bc52d...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang-nursery/rustup.rs/builds/266988652?utm_source=github_status&utm_medium=notification)

seanmonstar commented:
I had removed a `[replace]` for mio, as it was warning me that it wasn't needed. I'm not sure why, but I've added it back, to fix those tests on Android.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 2ac9fad has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 2ac9fadf96bb2c0edd2768e0c3733c13bd05a83a -->

bors commented:
:hourglass: Testing commit 2ac9fadf96bb2c0edd2768e0c3733c13bd05a83a with merge d1968e22a1512c229aa1990178ce3f35faa81986...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang-nursery/rustup.rs/builds/268082218?utm_source=github_status&utm_medium=notification)

seanmonstar commented:
I had included downgrades in the `Cargo.lock` that I didn't even notice. I've gone through and fixed every single one, besides the absolute necessary to needed in this PR.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 28f37ab has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 28f37ab0f8eeb4c1d21426bbcc063a0b1dfc2325 -->

bors commented:
:hourglass: Testing commit 28f37ab0f8eeb4c1d21426bbcc063a0b1dfc2325 with merge 1e191168ab5a152691720a97043241f2177cdeeb...

alexcrichton commented:
Tests passed, gonna merge manually.
